xarray/namedarray/core.py

```diff
@@ -0,0 +1,337 @@
+from __future__ import annotations
+
+import copy
+import itertools
+import math
+import typing
+from collections.abc import Hashable, Iterable
+from dataclasses import dataclass, field
+
+import numpy as np
+import numpy.typing as npt
+
+from xarray.namedarray.utils import Frozen, _default, is_duck_dask_array
+
+# temporary placeholder for indicating that an array api compliant
+# type. hopefully in the future we can narrow this down more
+T_Array = typing.TypeVar("T_Array", bound=typing.Any)
+
+
+@dataclass(slots=True)
+class NamedArray:
+    dims: tuple[Hashable, ...] = field(init=True)
+    data: npt.ArrayLike = field(init=True)
+    attrs: dict[typing.Any, typing.Any] = field(default_factory=dict, init=True)
+
+    _dims: tuple[Hashable, ...] = field(init=False)
+    _data: npt.ArrayLike = field(init=False)
+    _attrs: dict[typing.Any, typing.Any] = field(default_factory=dict, init=False)
+
+    def __post_init__(self):
+        self._data = self.data
```

Comment:
Missing `as_compatible_data`

Reply:
i've been avoiding using `as_compatible_data` ulitity function since it's full of xarray and pandas specific imports: 

for e.g.:

https://github.com/pydata/xarray/blob/83c2919b27b4b2d8a01bfa380226134c71321aa0/xarray/core/variable.py#L244-L260

curious to hear your thoughts about the alternative solutions

Reply:
Good point!

We'll need something similar for sure to make sure we have received an array. And to convert scalars at least